code/datums/status_effects/debuffs/blindness.dm

```diff
@@ -61,21 +67,47 @@
 	id = "blindness"
 	tick_interval = -1
 	alert_type = /atom/movable/screen/alert/status_effect/blind
+	var/static/list/update_signals = list(
+		SIGNAL_REMOVETRAIT(TRAIT_SIGHT_BYPASS),
+		SIGNAL_ADDTRAIT(TRAIT_SIGHT_BYPASS)
+	)
 	// This is not "remove on fullheal" as in practice,
 	// fullheal should instead remove all the sources and in turn cure this
 
 /datum/status_effect/grouped/blindness/on_apply()
 	if(!CAN_BE_BLIND(owner))
 		return FALSE
 
+
+	RegisterSignals(owner, update_signals,
+		PROC_REF(update_blindness))
+
+	update_blindness()
+
+	return ..()
+
+/datum/status_effect/grouped/blindness/proc/update_blindness()
+	if(!CAN_BE_BLIND(owner)) // future proofing
+		on_remove()
```

Comment:
As before don't call `on_remove` directly, this is asking for trouble

You probably want `qdel(src)`